pkg/cmd/roachtest/tests/mixed_version_decommission.go

```diff
@@ -12,328 +12,330 @@ package tests
 
 import (
 	"context"
+	gosql "database/sql"
 	"fmt"
-	"strconv"
+	"math/rand"
 	"time"
 
 	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster"
 	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/option"
 	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/roachtestutil"
 	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/roachtestutil/clusterupgrade"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/roachtestutil/mixedversion"
 	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test"
 	"github.com/cockroachdb/cockroach/pkg/roachprod/install"
+	"github.com/cockroachdb/cockroach/pkg/roachprod/logger"
 	"github.com/cockroachdb/cockroach/pkg/testutils"
-	"github.com/cockroachdb/cockroach/pkg/testutils/release"
-	"github.com/cockroachdb/cockroach/pkg/util/httputil"
-	"github.com/cockroachdb/cockroach/pkg/util/retry"
-	"github.com/cockroachdb/cockroach/pkg/util/version"
 	"github.com/cockroachdb/errors"
 )
 
-// runDecommissionMixedVersions runs through randomized
-// decommission/recommission processes in mixed-version clusters.
-func runDecommissionMixedVersions(
-	ctx context.Context, t test.Test, c cluster.Cluster, buildVersion *version.Version,
-) {
-	predecessorVersionStr, err := release.LatestPredecessor(buildVersion)
-	if err != nil {
-		t.Fatal(err)
-	}
-	predecessorVersion := clusterupgrade.MustParseVersion(predecessorVersionStr)
-
-	h := newDecommTestHelper(t, c)
-
-	pinnedUpgrade := h.getRandNode()
-	t.L().Printf("pinned n%d for upgrade", pinnedUpgrade)
-
+func runDecommissionMixedVersions(ctx context.Context, t test.Test, c cluster.Cluster) {
 	// NB: The suspect duration must be at least 10s, as versions 23.2 and
 	// beyond will reset to the default of 30s if it fails validation, even if
 	// set by a previous version.
 	const suspectDuration = 10 * time.Second
 
-	allNodes := c.All()
-	u := newVersionUpgradeTest(c,
-		// We upload both binaries to each node, to be able to vary the binary
-		// used when issuing `cockroach node` subcommands.
-		uploadCockroachStep(allNodes, predecessorVersion),
-		uploadCockroachStep(allNodes, clusterupgrade.CurrentVersion()),
-
-		startVersion(allNodes, predecessorVersion),
-		waitForUpgradeStep(allNodes),
-		preventAutoUpgradeStep(h.nodeIDs[0]),
-		suspectLivenessSettingsStep(h.nodeIDs[0], suspectDuration),
-
-		preloadDataStep(pinnedUpgrade),
-
-		// We upgrade a pinned node and one other random node of the cluster to the current version.
-		binaryUpgradeStep(c.Node(pinnedUpgrade), clusterupgrade.CurrentVersion()),
-		binaryUpgradeStep(c.Node(h.getRandNodeOtherThan(pinnedUpgrade)), clusterupgrade.CurrentVersion()),
-		checkAllMembership(pinnedUpgrade, "active"),
-
-		// After upgrading, which restarts the nodes, ensure that nodes are not
-		// considered suspect unnecessarily.
-		sleepStep(2*suspectDuration),
-
-		// Partially decommission a random node from another random node. We
```

Comment:
What was reasoning for getting rid of the randomness in node choice here? It also means we won't even do a self decommission in this test right?

Reply:
Mostly because the randomness is now "built-in". In each run of this test, the versions running on n1 and n2 could be any combination of `{old, new} x {old, new}`.

Good catch on self decommission, I added that in.